### PR TITLE
Hallucinations and tokens

### PR DESCRIPTION
**Pull request by EzraEllette**
---
name: pull request
about: submit changes to the project
title: "[pr] "
labels: ''
assignees: ''

---

## description
remove timestamp tokens and obvious hallucinations from speech to text

### CODE REVIEW

File: screenpipe-audio/src/vad_engine.rs [hunk 1]
```diff
@@ -225,6 +249,34 @@ impl VadEngine for SileroVad {
         Ok(status == VadStatus::Speech && result.prob > threshold)
     }
 
+    fn audio_type(&mut self, audio_chunk: &[f32]) -> anyhow::Result<VadStatus> {
```
Comment (louis030195) [inline screenpipe-audio/src/vad_engine.rs:252]: @EzraEllette fyi we might want to remove webrtc altogether, which does not make sense to use anyway and is causing issues on some OS

https://github.com/mediar-ai/screenpipe/issues/442

File: screenpipe-audio/src/stt.rs [hunk 2]
```diff
@@ -367,11 +375,59 @@ pub async fn stt(
...
+
+            let mut min_time: f32 = f32::MAX;
+            let mut max_time: f32 = f32::MIN;
+            let mut i = 0;
+            let segments_len = segments.len();
+            for segment in segments {
+                let mut text = segment.dr.text.clone();
+
+                // maybe not <|0.00|> but <|12.34|>
+                let mut start = text[..8].to_string();
```
Comment (louis030195) [inline screenpipe-audio/src/stt.rs:391]: can you explain this part in plain english?

also i think we should really use overlap somehow

example:

bob talking for 45 seconds

screenpipe record 30 then 30 and whisper will lose quality because not having context from the previous audio chunk

wdyt?

**Comment by louis030195** [comment]
from what i understand this pr takes this changes in https://github.com/mediar-ai/screenpipe/pull/459

happy to merge asap to provide higher quality now to everyone - giving you part of the bounty for now 

/tip $200 @EzraEllette

**Comment by EzraEllette** [comment]
Thanks Louis! This is a combination of Spectral Subtraction and Token Cleanup. It's not perfect with hallucination catching, but It is a big improvement

**Comment by louis030195** [comment]
@EzraEllette nice, just tested, some error showed up at some point:

Screenshot 2024-10-11 at 15 39 05
Screenshot 2024-10-11 at 15 38 44
Screenshot 2024-10-11 at 15 38 01

also repeatitive? but seems issue was just on the first 60s of start, also noticed russian but maybe it's not a new bug 

related: #469

**Comment by EzraEllette** [comment]
I'll take a look at this with some longer audio samples to solve the issue and incorporate overlapping in a separate PR.

**Comment by EzraEllette** [comment]
Can you share the recording that's causing an issue?

File: screenpipe-audio/src/stt.rs (hunk 2 above)
Comment (EzraEllette) [reply screenpipe-audio/src/stt.rs:391]: I'm pretty busy right now I'm commenting to confirm that I see this. I'll answer you and add some comments to the code tomorrow.

**Comment by louis030195** [comment]
https://www.youtube.com/results?search_query=steve+jobs+speech

i just started screenpipe on your branch with this

**Comment by EzraEllette** [comment]
image
The repetition is present on the main branch as well. Checking to see if i can get random language detections.

**Comment by EzraEllette** [comment]
image

**Comment by EzraEllette** [comment]
I think a PR is necessary to make whisper-large-v3-turbo the default transcription engine. I was getting the repetition with distill, but not turbo. Have you noticed the same @louis030195 ?

**Comment by EzraEllette** [comment]
overlaps:

```
Second story about love and loss. I was lucky. I found what I loved to do early in life. Waz and I started Apple in my parents' garage when I was 20. We worked hard and in 10 years Apple had grown from just the two of us in a garage into a $2 billion company with over 4,000 employees. We had just released our finest creation, the Macintosh, a year earlier and I had just turned 30 and I got fired.


I got fired. How can you get fired from a company you started? Well, as Apple grew, we hired someone who I thought was very talented to run the company with me. And for the first year or so, things went well. But then our visions of the future began to diverge, and eventually we had a falling out. When we did, our board of directors sided with him. And so at 30, I was out, and very publicly out. Then the focus of my entire adult life was to be a part of the company.


had been the focus of my entire adult life was gone and it was devastating. I didn't know what to do for a few months. I felt that I'd let the previous generation of entrepreneurs down, that I had dropped the baton as it was being passed to me. I met with David Packard and Bob Noyce and tried to apologize for screwing up so badly. It was a very public failure and I even thought about running away from the valley. Something slowly began to dawn on me. I still loved what I did.


I still loved what I did. The turn of events at Apple had not changed that one bit. I had been rejected, but I was still in love. So I decided to start over. I didn't see it then, but it turned out that getting fired from Apple was the best thing that could have ever happened to me. The heaviness of being successful was replaced by the lightness of being a beginner again, less sure about everything. It freed me to enter one of the most creative periods of my life.


periods of my life. The next five years I started a company named Next, another company named Pixar and fell in love with an amazing woman who would become my wife. Pixar went on to create the world's first computer animated feature film, Toy Story, and is now the most successful animation studio in the world. A remarkable turn of events, Apple bought Next and I returned to Apple and the technology we developed at Next is at the heart of Apple's current Renaissance.


Apple's current renaissance and Lorene and I have a wonderful family together. I'm pretty sure none of this would have happened if I hadn't been fired from Apple. It was awful tasting medicine, but I guess the patient needed it. Sometimes life's going to hit you in the head with a brick. Don't lose faith. I'm convinced that the only thing that kept me going was that I loved what I did. You've got to find what you love and that is as true for work as it is for your lovers.


Work as it is for your lovers. Work is going to fill a large part of your life and the only way to be truly satisfied is to do what you believe is great work. And the only way to do great work is to love what you do. If you haven't found it yet, keep looking and don't settle. As with all matters of the heart, you'll know when you find it. And like any great relationship, it just gets better and better as the years roll on. So keep looking, don't settle.
```

**Comment by EzraEllette** [comment]
I think one of the biggest problems with the recording implementation is that there is a delay between the recordings. This can be fixed with the overlap, but I also think about an implementation where audio never stops recording, but every X seconds, the audio buffer is sent to the whisper_sender channel and cleared.

**Comment by louis030195** [comment]
> I think one of the biggest problems with the recording implementation is that there is a delay between the recordings. This can be fixed with the overlap, but I also think about an implementation where audio never stops recording, but every X seconds, the audio buffer is sent to the whisper_sender channel and cleared.

make sense 

will test pr again and merge if the issue was distil model

also some additional context #170 trying to find how to properly fit this in the future (always record default devices i guess)
